xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp

```diff
@@ -1928,92 +1920,142 @@ void COutput::InitCycle()
-        else if (method == VS_INTERLACEMETHOD_VAAPI_MADI)
-          m_config.processInfo->SetVideoDeintMethod("vaapi-madi");
-        else if (method == VS_INTERLACEMETHOD_VAAPI_MACI)
-          m_config.processInfo->SetVideoDeintMethod("vaapi-mcdi");
+        switch (method)
+        {
+          case VS_INTERLACEMETHOD_DEINTERLACE:
+            m_config.processInfo->SetVideoDeintMethod("yadif");
+            break;
+          case VS_INTERLACEMETHOD_RENDER_BOB:
+            m_config.processInfo->SetVideoDeintMethod("render-bob");
+            break;
+          case VS_INTERLACEMETHOD_VAAPI_BOB:
+            m_config.processInfo->SetVideoDeintMethod("vaapi-bob");
+            break;
+          case VS_INTERLACEMETHOD_VAAPI_MADI:
+            m_config.processInfo->SetVideoDeintMethod("vaapi-madi");
+            break;
+          case VS_INTERLACEMETHOD_VAAPI_MACI:
+            m_config.processInfo->SetVideoDeintMethod("vaapi-mcdi");
+            break;
+          default:
+            m_config.processInfo->SetVideoDeintMethod("unknown");
+            break;
+        }
       }
       else
       {
+        CLog::Log(LOGERROR, "VAAPI output: Postproc preinit failed");
         delete m_pp;
-        m_pp = NULL;
+        m_pp = nullptr;
       }
     }
   }
   // progressive
   else
   {
     method = VS_INTERLACEMETHOD_NONE;
+
     if (m_pp && !m_pp->UpdateDeintMethod(method))
     {
+      CLog::Log(LOGDEBUG, LOGVIDEO, "VAAPI output: Current postproc does not want new deinterlace mode, removing");
       delete m_pp;
-      m_pp = NULL;
+      m_pp = nullptr;
       DropVppProcessedPictures();
     }
+
     if (!m_pp)
     {
+      const bool preferVaapiRender = CServiceBroker::GetSettings().GetBool(CSettings::SETTING_VIDEOPLAYER_PREFERVAAPIRENDER);
+      // For 1080p/i or below, always use CVppPostproc even when not deinterlacing
+      // Reason is: mesa cannot dynamically switch surfaces between use for VAAPI post-processing
+      // and use for direct export, so we run into trouble if we or the user want to switch
+      // deinterlacing on/off mid-stream.
+      // See also: https://bugs.freedesktop.org/show_bug.cgi?id=105145
+      const bool alwaysInsertVpp = (m_config.vidWidth * m_config.vidHeight) <= (1920 * 1080);
+
       m_config.stats->SetVpp(false);
-      if (!CServiceBroker::GetSettings().GetBool(CSettings::SETTING_VIDEOPLAYER_PREFERVAAPIRENDER))
+      if (!preferVaapiRender)
+      {
+        CLog::Log(LOGINFO, LOGVIDEO, "VAAPI output: Initializing ffmpeg postproc");
         m_pp = new CFFmpegPostproc();
+      }
+      else if (alwaysInsertVpp)
+      {
+        CLog::Log(LOGINFO, LOGVIDEO, "VAAPI output: Initializing vaapi postproc");
+        m_pp = new CVppPostproc();
+        m_config.stats->SetVpp(true);
+      }
       else
       {
+        CLog::Log(LOGINFO, LOGVIDEO, "VAAPI output: Initializing skip postproc");
         m_pp = new CSkipPostproc();
-        m_config.stats->SetVpp(true);
```

Comment:
Actually - what meaning does `SetVpp` have? Should it be true for SkipPostproc?

Reply:
ok, does not have any function atm